tools/benchmark/src/utilities.ts

```diff
@@ -0,0 +1,189 @@
+	/**
+	 * Margin of error.
+	 */
+	readonly marginOfError: number;
+
+	/**
+	 * Margin of error as a percentage of the mean.
+	 */
+	readonly marginOfErrorPercent: number;
+
+	/**
+	 * Standard error of the mean.
+	 */
+	readonly standardErrorOfMean: number;
+
+	/**
+	 * Standard deviation.
+	 */
+	readonly standardDeviation: number;
+
+	/**
+	 * Arithmetic mean.
+	 */
+	readonly arithmeticMean: number;
+
+	/**
+	 * Sample values.
+	 */
+	readonly samples: readonly number[];
+
+	/**
+	 * Variance.
+	 */
+	readonly variance: number;
+}
+
+/**
+ * Contains the samples of all memory-related measurements we track for a given benchmark (a test which was
+ * potentially iterated several times). Each property is an array and all should be the same length, which
+ * is the number of iterations done during the benchmark.
+ * @public
+ */
+export interface MemoryTestData {
+	/**
+	 * Memory usage in bytes.
+	 */
+	memoryUsage: NodeJS.MemoryUsage[];
+
+	/**
+	 * Heap info.
+	 */
+	heap: v8.HeapInfo[];
+
+	/**
+	 * Heap space info.
+	 */
+	heapSpace: v8.HeapSpaceInfo[][];
+}
+
+/**
+ * Result of successfully running a benchmark.
+ * @public
+ */
+export interface BenchmarkData {
+	/**
+	 * Iterations per batch.
+	 */
+	readonly iterationsPerBatch: number;
+
+	/**
+	 * Number of batches, each with `iterationsPerBatch` iterations.
+	 */
+	readonly numberOfBatches: number;
+
+	/**
+	 * Stats about runtime, in seconds.
+	 * This is already scaled to be per iteration and not per batch.
+	 */
+	readonly stats: Stats;
+
+	/**
+	 * Time it took to run the benchmark in seconds.
+	 */
+	readonly elapsedSeconds: number;
+}
+
+/**
+ * Result of failing to run a benchmark.
+ * @public
+ */
+export interface BenchmarkError {
+	/**
+	 * Error message.
+	 */
+	error: string;
+}
+
+/**
+ * Contains the full results for a benchmark (a test which was potentially iterated several times).
+ * 'samples' contains the raw 'before' and 'after' measurements instead of calculated deltas
+ * for flexibility, at the cost of more memory and bigger output.
+ * @public
+ */
+export interface MemoryBenchmarkStats {
+	/**
+	 * Number of runs.
+	 */
+	runs: number;
+
+	/**
+	 * Contains the samples of all memory-related measurements we track for a given benchmark (a test which was
+	 * potentially iterated several times). Each property is an array and all should be the same length, which
+	 * is the number of iterations done during the benchmark.
+	 */
+	samples: { before: MemoryTestData; after: MemoryTestData };
+
+	/**
+	 * Stats about the arithmetic mean of the delta between the 'before' and 'after' measurements.
+	 */
+	stats: Stats;
+
+	/**
+	 * Whether the benchmark was aborted.
+	 */
+	aborted: boolean;
+
+	/**
+	 * Error message.
+	 */
+	error?: Error;
+
+	/**
+	 * Total run time in milliseconds.
+	 */
+	totalRunTimeMs: number;
+}
+
+/**
+ * Result of trying to run a benchmark.
+ * @public
+ */
+export type BenchmarkResult = BenchmarkError | BenchmarkData | MemoryBenchmarkStats;
```

Comment:
I think the approach is still weird here, causing us to expose separate kinds of objects for runtime benchmarks and for memory benchmarks. The idea is to consolidate on a single type of object that can have arbitrary properties, so each kind of benchmark can put whatever it needs there, and we wouldn't need separate types.

What I had in mind is that BenchmarkResult can be either BenchmarkError or BenchmarkData. BenchmarkData would have common properties that apply to all kinds of benchmark (e.g. iterationsPerBatch, numberOfBatches, elapsedSeconds) and one extra property `customData: Record<string, unknown>`. The structure of `Stats` is the customData for a runtime benchmark. The structure of MemoryBenchmarkStats is the customData for a memory benchmark. (Maybe both of them need a bit of tweaking, like removing the samples). But we wouldn't want them to be types on their own; or at least, they shouldn't be _common_ types, they could be types defined in the corresponding runner if they help to write that code, and the runners should just produce a BenchmarkResult (which is either a BenchmarkError if something went wrong with the test, or a BenchmarkData, with the common properties, and whatever is specific to that particular runner inside the benchmark result's `customData`).

In that world, the reporter code does not have any need for functions like `isMemoryBenchmarkStats`. It only ever cares about BenchmarkResult, it knows about its common properties and does what it needs with them, and then whatever comes in the result's `customData`, the reporter just uses blindly as key-value pairs. Also in that world the MochaMemoryTestReporter is not necessary anymore and can be completely removed, so I think we should see that in this PR as well.

Reply:
Updates:
1. Remove `MemoryBenchmarkStats` and add `customData: Record<string, unknown>` in `BenchmarkData`.
2. Move fields not in `BenchmarkData` into `customData`
3. Remove all isType function
4. Update `MemoryTestRunner` to use `BenchmarkResult`, all new runner will only need to use `BenchmarkResult`